spirv_glsl.cpp

```diff
@@ -10239,7 +10244,14 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
 		// If an expression is mutable and forwardable, we speculate that it is immutable.
 		AccessChainMeta meta;
 		bool ptr_chain = opcode == OpPtrAccessChain;
-		auto e = access_chain(ops[2], &ops[3], length - 3, get<SPIRType>(ops[0]), &meta, ptr_chain);
+		auto target_type = get<SPIRType>(ops[0]);
+		auto e = access_chain(ops[2], &ops[3], length - 3, target_type, &meta, ptr_chain);
+
+		if (should_emit_temporary(ops[2], target_type))
+		{
+			statement(declare_temporary(ops[0], ops[1]), move(e), ";");
```

Comment:
This is incorrect. The more correct approach here is to block any forwarding in the SPIRExpression instead (see last argument). You can use OpCompositeConstruct as a guideline here.

Reply:
Do you mean that the statement shouldn't be emitted from OpAccessChain? I have ever tried making `SPIRExpression::immutable` `true` and that results emitting not stuffs like `Light _51 = {UBO[_96 * 2 + 4].xyz, UBO[_96 * 2 + 4].w, UBO[_96 * 2 + 5]};` but like `float3 _101 = ({UBO[_96 * 2 + 4].xyz, UBO[_96 * 2 + 4].w, UBO[_96 * 2 + 5]}).Position;`. The OpCompositeConstruct case in `CompilerGLSL::emit_instruction` is calling `emit_op`, and `emit_op` calls `statement` if the given `forwarding` is `false`. It looks not using the last argument of `SPIRExpression` to me. I didn't understand what I should refer to.

Reply:
Right, emit_op() is probably the thing to call here. This is so that end up binding the expression to a temporary, which is the only way HLSL can generate struct composites.